---
title: "Dragen-vs-bcbio variant evaluation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("evaluation.R")
```

Comparing tumor-normal small somatic variant calling performed by Dragen vs [bcbio-nextgen](https://bcbio-nextgen.readthedocs.org) (BWA, and ensemble 2 out of 3 callers: Strelka2, Mutect2, and Vardict). The input data is COLO829 cell line, a dilution seriues of 40%, 60%, 80%, and 100% tumor purities, made by artificially mixing blood and tumor samples. As a reference truth set, using the one produced by [Craig et al](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4837349/).

Dragen was run using the following arguments:

```
/opt/edico/bin/dragen --output-directory /newvolume/colo100pc_align/ --tumor-fastq1 Colo829_S5_R1_001.fastq.gz --tumor-fastq2 Colo829_S5_R2_001.fastq.gz -1 Colo829_B_S16_R1_001.fastq.gz -2 Colo829_B_S16_R2_001.fastq.gz --output-file-prefix Colo829_100pc --ref-dir /staging/hg38 --enable-variant-caller true --enable-map-align true --vc-sample-name Colo829_100pc --output-format BAM --enable-map-align-output true
```

Variants produced by both pipelines were additionally filtered using the [umccrise approach](https://github.com/umccr/umccrise/blob/master/workflow.md#somatic) (excluding SAGE part).



TODO: try comparing before filtering
TODO: try truth before filtering


```{r}
dir = "/Users/vsaveliev/Analysis/Dragen/"

truth_vcf = read.vcf(str_c(dir, "hg38/truth_onesample.hg38.PREP.NOALT.ANNO.FILT.vcf.gz"), split.info = T)

# dragen100_vcf = read.vcf(str_c(dir, "hg38/dragen/Colo829_100pc.PREP.NOALT.ANNO.FILT.vcf.gz"), split.info = T)
# dragen80_vcf  = read.vcf(str_c(dir, "hg38/dragen/Colo829_80pc.PREP.NOALT.ANNO.FILT.vcf.gz"),  split.info = T)
# dragen60_vcf  = read.vcf(str_c(dir, "hg38/dragen/Colo829_60pc.PREP.NOALT.ANNO.FILT.vcf.gz"),  split.info = T)
# dragen40_vcf  = read.vcf(str_c(dir, "hg38/dragen/Colo829_40pc.PREP.NOALT.ANNO.FILT.vcf.gz"),  split.info = T)
# 
# strelka100_vcf = read.vcf(str_c(dir, "hg38/strelka/Colo829_100pc.ANNO.FILT.vcf.gz"), split.info = T)
# strelka80_vcf  = read.vcf(str_c(dir, "hg38/strelka/Colo829_80pc.ANNO.FILT.vcf.gz"),  split.info = T)
# strelka60_vcf  = read.vcf(str_c(dir, "hg38/strelka/Colo829_60pc.ANNO.FILT.vcf.gz"),  split.info = T)
# strelka40_vcf  = read.vcf(str_c(dir, "hg38/strelka/Colo829_40pc.ANNO.FILT.vcf.gz"),  split.info = T)
#  
# dragen100_eval = merge_called_and_truth(dragen100_vcf, truth_vcf) %>% mutate(TLOD = as.double(TLOD), NLOD = as.double(NLOD), purity = 100)
# dragen80_eval  = merge_called_and_truth(dragen80_vcf, truth_vcf)  %>% mutate(TLOD = as.double(TLOD), NLOD = as.double(NLOD), purity = 80)
# dragen60_eval  = merge_called_and_truth(dragen60_vcf, truth_vcf)  %>% mutate(TLOD = as.double(TLOD), NLOD = as.double(NLOD), purity = 60)
# dragen40_eval  = merge_called_and_truth(dragen40_vcf, truth_vcf)  %>% mutate(TLOD = as.double(TLOD), NLOD = as.double(NLOD), purity = 40)

bwa = tibble()
dragen = tibble()
dragen_strelka = tibble()
for (p in c(20, 40, 60, 80, 100)) {
  # for (c in c("strelka2", "ensemble")) {
  #   vcf = read.vcf(str_c(dir, "hg38/bcbio_bwa/Colo829_", as.character(p), "pc_", c, "_bwa.vcf.gz"), split.info = T)
  #   loaded = merge_called_and_truth(vcf, truth_vcf) %>% mutate(caller = str_c("bwa_", c), purity = p)
  #   bwa = bind_rows(bwa, loaded)
  # }
  
  vcf = read.vcf(str_c(dir, "hg38/dragen/Colo829_", as.character(p), "pc.PREP.NOALT.ANNO.FILT.vcf.gz"), split.info = T)
  loaded = 
    merge_called_and_truth(vcf, truth_vcf) %>%
    mutate(TLOD = as.double(TLOD), 
           NLOD = as.double(NLOD), 
           purity = p,
           caller = "dragen_dragen")
  dragen = bind_rows(dragen, loaded)
  
  # vcf = read.vcf(str_c(dir, "hg38/strelka/Colo829_", as.character(p), "pc.ANNO.FILT.vcf.gz"), split.info = T)
  # loaded = merge_called_and_truth(vcf, truth_vcf) %>% 
  #   mutate(purity = p,
  #          caller = "dragen_strelka")
  # dragen_strelka = bind_rows(dragen_strelka, loaded)
}

# dragen = bind_rows(
#   dragen100_eval %>% mutate(purity = 100),
#   dragen80_eval  %>% mutate(purity = 80),
#   dragen60_eval  %>% mutate(purity = 60),
#   dragen40_eval  %>% mutate(purity = 40)
# ) %>% 
#   mutate(caller = "dragen_dragen") %>% 
#   rescue_filt("LowTLOD")
# 
# # bwa %>% filter(caller == "bwa_ensemble") %>% filter(is_fn) %>% count(is.na(AF))
data = bind_rows(
    dragen_strelka,
    dragen,
    bwa
  ) %>% mutate(is_true = is_true & FILT.truth == 'PASS') %>% count_status()    
```

Comparing filtered/unfiltered

```{r}
# mutate(caller = str_c(caller, '_60pc')) %>% 
# data %>% 
#   filter(purity == 60) %>% 
#   filter(is_passed | is_true) %>% 
#   filter(!is.na(vartype)) %>% 
#   ggplot() + 
#   geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
#   facet_wrap(~status, nrow = 3, scales = 'free_y') +
#   scale_x_continuous(breaks = seq(0, 1, 0.1)) +
#   theme_grey(base_size = 16)
# 
# data_unfiltered = data %>% 
#   mutate(
#     umccrise_passed = is_called & AF > 0.1,
#     is_passed = umccrise_passed
#   ) %>%
#   count_status()
# 
# all = bind_rows(data %>% mutate(filtering = 'default'), 
#                 data_unfiltered %>% mutate(filtering = "unfiltered_calls"))
```

```{r fig.width = 15}
# all = all %>% filter(purity == 60) %>% filter(is_passed | is_true) %>% filter(!is.na(vartype))
# d1 = all %>% filter(filtering == 'default') 
# d2 = all %>% filter(filtering == 'unfiltered_calls')
# library(patchwork)
# p1 = ggplot(d1) + geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) + facet_wrap(~status, nrow = 1, scales = "free") + scale_x_continuous(breaks = seq(0, 1, 0.1))
# p2 = ggplot(d2) + geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) + facet_wrap(~status, nrow = 1, scales = "free") + scale_x_continuous(breaks = seq(0, 1, 0.1)) 
# p1 / p2

#################
## unfiltered with tlods
data_unfiltered = data %>% 
  mutate(
    umccrise_passed = is_called & AF >= 0.1,
    is_passed = umccrise_passed
  ) %>%
  count_status()

dragen_unfiltered = data_unfiltered %>% filter(caller == "dragen_dragen")

with_tlods_unfiltered = bind_rows(
  dragen_unfiltered,
  dragen_unfiltered %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_dragen_tlod21'),
  dragen_unfiltered %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_dragen_tlod15'),
  data_unfiltered %>% filter(caller == "bwa_ensemble" | caller == "bwa_strelka2")
)

##################
## filtered with tlods
dragen = data %>% filter(caller == "dragen_dragen")
dragen = dragen %>% rescue_filt("LowTLOD")
with_tlods = bind_rows(
  dragen,
  dragen %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_dragen_tlod21'),
  dragen %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_dragen_tlod15'),
  data %>% filter(caller == "bwa_ensemble" | caller == "bwa_strelka2")
)

#################
## compare filtered vs unfiltered
filtered_vs_unfiltered = bind_rows(
  with_tlods %>% mutate(filtering = "umccrise"), 
  with_tlods_unfiltered %>% mutate(filtering = "AF>10%")
)

filtered_vs_unfiltered %>% 
  filter(purity == 60) %>% 
  mutate(caller = factor(caller, levels = c("bwa_ensemble", "dragen_dragen", "dragen_dragen_tlod15", "dragen_dragen_tlod21"))) %>% 
  filter(is_passed | is_true) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(status ~ filtering, ncol = 2, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1)) 

with_tlods %>% filter(caller == "bwa_ensemble") %>% count(is_passed)
with_tlods_unfiltered %>% filter(caller == "bwa_ensemble") %>% count(is_passed)
```

```{r fig.width = 15}
toplot = filtered_vs_unfiltered %>% 
  mutate(caller = factor(caller, levels = c("bwa_ensemble", "dragen_dragen", "dragen_dragen_tlod15", "dragen_dragen_tlod21"))) %>% 
  # filter(caller == "bwa_ensemble") %>% 
  filter(is_passed | is_true) %>% 
  filter(AF <= 0.9, AF >= 0.1)

fn = toplot %>% filter(is_fn)
fp = toplot %>% filter(is_fp)
tp = toplot %>% filter(is_tp)
(fn %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), linetype = filtering, color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) 
) / 
(fp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), linetype = filtering, color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) 
) / 
(tp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), linetype = filtering, color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) 
)
```

```{r fig.width = 8}
unfiltered_ensemble = bind_rows(
  # with_tlods_unfiltered %>% filter(caller == "bwa_ensemble") %>% mutate(caller = 'bcbio: bwa+ensemble'), 
  # with_tlods_unfiltered %>% filter(caller == "bwa_strelka2") %>% mutate(caller = 'bcbio: bwa+strelka2'),
  with_tlods_unfiltered %>% filter(caller == "dragen_dragen") %>% mutate(caller = 'dragen'),
  with_tlods %>% filter(caller == "dragen_dragen") %>% mutate(caller = "dragen, filtered with umccrise"),
  # data_unfiltered %>% filter(caller == "dragen_strelka") %>% mutate(caller = "dragen+strelka2"),
  with_tlods %>% filter(caller == "dragen_dragen_tlod15") %>% mutate(caller = "dragen, filtered with umccrise + tlod>=15")
  # with_tlods %>% filter(caller == "dragen_dragen_tlod21") %>% mutate(caller = "dragen, filtered with umccrise + tlod>=21")
)



toplot = unfiltered_ensemble %>% 
  mutate(caller = factor(caller, levels = c("bcbio: bwa+ensemble", 
                                            "dragen", 
                                            "dragen, filtered with umccrise", 
                                            "dragen+strelka2", 
                                            "dragen, filtered with umccrise + tlod>=15",
                                            "bcbio: bwa+strelka2"
                                            ))) %>% 
  filter(is_passed | is_true)  
  # filter(AF <= 0.9, AF >= 0.1)
  # filter(caller == "dragen+strelka2" | caller == "bcbio: bwa+strelka2")

# toplot2 = toplot %>% 
  # mutate(vartype = factor(ifelse(vartype == 1, "SNP", "Indel"), levels = c("SNP", "Indel")))

# toplot %>% filter(is_tp) %>% 
#   ggplot() + 
#   geom_freqpoly(aes(AF, ..count.., color = caller), binwidth = 0.05) +
#   facet_wrap(vartype ~ purity, nrow = 2, scales = 'free_y') +
#   scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
#   labs(title = "True positives. Left to right: tumor purity percentage")

(toplot %>% filter(is_tp) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, ..count.., color = caller), binwidth = 0.05) +
  facet_wrap(vartype ~ purity, nrow = 2, scales = 'free_y') +
  # scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "True positives. Left to right: tumor purity percentage")
) /
(toplot %>% filter(is_fp) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, ..count.., color = caller), binwidth = 0.05) +
  facet_wrap(vartype ~ purity, nrow = 2, scales = 'free_y') +
  # scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "False positives")
) / 
(toplot %>% filter(is_fn) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, ..count.., color = caller), binwidth = 0.05) +
  facet_wrap(vartype ~ purity, nrow = 2, scales = 'free_y') +
  # scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "False negatives (missed variants)")
) 
```

```{r}
be = with_tlods_unfiltered %>% filter(caller == "bwa_ensemble") %>% mutate(caller = 'bcbio: bwa+ensemble')          %>% filter(purity == 80)
bs =  with_tlods_unfiltered %>% filter(caller == "bwa_strelka2") %>% mutate(caller = 'bcbio: bwa+strelka2')           %>% filter(purity == 80)
d =  with_tlods_unfiltered %>% filter(caller == "dragen_dragen") %>% mutate(caller = 'dragen')                    %>% filter(purity == 80)
du =  with_tlods %>% filter(caller == "dragen_dragen") %>% mutate(caller = "dragen, filtered with umccrise")             %>% filter(purity == 80)
ds =  data_unfiltered %>% filter(caller == "dragen_strelka") %>% mutate(caller = "dragen+strelka2")                    %>% filter(purity == 80)
du15 =  with_tlods %>% filter(caller == "dragen_dragen_tlod15") %>% mutate(caller = "dragen, filtered with umccrise + tlod>=15")%>% filter(purity == 80)
show_stats(bs, ds, d, du, du15, be) 
```

```{r}
with_tlods %>% distinct(vartype)

filtered = with_tlods %>% filter(caller == "dragen_dragen")  # %>%  mutate(vartype = ifelse(vartype == 1, "SNP", "Indel"))
unfiltered = with_tlods_unfiltered %>% filter(caller == "dragen_dragen")  # %>%  mutate(vartype = ifelse(vartype == 1, "SNP", "Indel"))
p20_tlod15 = filtered %>% filter(purity == 20)
p40_tlod15 = filtered %>% filter(purity == 40)
p60_tlod15 = filtered %>% filter(purity == 60)
p80_tlod15 = filtered %>% filter(purity == 80)
p100_tlod15 = filtered %>% filter(purity == 100)
p20 = unfiltered %>% filter(purity == 20)
p40 = unfiltered %>% filter(purity == 40)
p60 = unfiltered %>% filter(purity == 60)
p80 = unfiltered %>% filter(purity == 80)
p100 = unfiltered %>% filter(purity == 100)
show_stats(p20_tlod15, p40_tlod15, p60_tlod15, p80_tlod15, p100_tlod15, p20, p40, p60, p80, p100) 


```

```{r fig.width = 12}
fn = toplot %>% filter(is_fn)
fp = toplot %>% filter(is_fp)
tp = toplot %>% filter(is_tp)
(tp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "True positives. Left to right: tumor purity percentage")
) /
(fp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "False positives")
) / 
(fn %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "False negatives (missed variants)")
) 
```


```{r}
data %>% 
  filter(purity == 60) %>% 
  filter(is_passed | is_true) %>% 
  filter(!is.na(vartype)) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(~ status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  theme_grey(base_size = 16)
```

```{r}
data_unfiltered_truthfilt %>% 
  filter(purity == 60) %>% 
  filter(is_passed | is_true) %>% 
  filter(!is.na(vartype)) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(~status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1)) +
  theme_grey(base_size = 16)
```

```{r fig.width=7}
bind_rows(
  dragen,
  dragen %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_tlod21'),
  dragen %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_tlod15')
) %>% 
filter(is_passed | is_true) %>% 
filter(!is.na(vartype)) %>% 
# filter(purity == 60) %>% 
ggplot() + 
geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
facet_wrap(status~purity, ncol=4, scales = 'free_y') +
scale_x_continuous(breaks = seq(0, 1, 0.1))
```

TP charts look suspiciously even for different TLOD thresholds. Checking if it's not a bug by trying few more extreme thresholds. All good:

```{r fix.width=15}
bind_rows(
  dragen,
  dragen %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_tlod21'),
  dragen %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_tlod15'),
  dragen %>% reject_if(TLOD < 30) %>% mutate(caller = 'dragen_tlod30'),
  dragen %>% reject_if(TLOD < 100) %>% mutate(caller = 'dragen_tlod100'),
  dragen %>% reject_if(TLOD < 200) %>% mutate(caller = 'dragen_tlod200')
  ) %>% 
  filter(is_tp) %>% 
  filter(purity == 60) %>%
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

Adding Strelka2. Big piles of FP on low frequency. Perhaps it has to do with Strelka2 not optimized to Dragen alignments. Should try with BWA.

Also wondering about this peak for 100% purity at 100% AF. Is it a problem with T/N calling?

Adding now BWA/ensemble and BWA/strelka.

```{r fig.width=7}
with_tlods = bind_rows(
  dragen %>% mutate(caller = "dragen_dragen"),
  dragen %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_dragen_tlod21'),
  dragen %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_dragen_tlod15'),
  strelka %>% mutate(caller = "dragen_strelka2"),
  data %>% filter(caller == "bwa_strelka2"),
  data %>% filter(caller == "bwa_ensemble")
) 

with_tlods %>% filter(caller == "bwa_ensemble") %>% filter(is_fn) %>% count(AF)

with_tlods %>% 
  mutate(
    caller = factor(caller, levels = c("bwa_ensemble", "dragen_dragen", "dragen_dragen_tlod15", "dragen_dragen_tlod21", "dragen_strelka2", "bwa_strelka2"))
  ) %>% 
  filter(is_passed | is_true) %>% 
  # filter(purity == 60) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(status ~ purity, ncol=4, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

Clearly dragen/dragen and bwa/ensemble are the best combinations. All other mixtures underperform. On on hand, this is expected; on the other, PURPLE and GRIDSS might underperform too running from Dragen alignments.

Now comparing bwa/ensemble vs. dragen/dragen specifically.

```{r fig.width=7}
dragen_bwa = bind_rows(
  dragen %>% mutate(caller = "dragen_dragen"),
  dragen %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_dragen_tlod21'),
  dragen %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_dragen_tlod15'),
  data %>% filter(caller == "bwa_ensemble")
)
dragen_bwa %>% 
  filter(is_passed | is_true) %>% 
  # filter(purity == 60) %>% 
  # filter(caller == "bwa_ensemble") %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(status~purity, ncol=4, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

Exploring different TLOD thresholds. The threshold around 14 is optimal for purities 60-100, and no filtering is optimal for 40, however 14 looks good anyway.

```{r, fig.width=12}
tlods = c(10:30)
purities = dragen %>% distinct(purity)
bytlods = tlods %>% map_df(function(tlod) {
    # tlod = 5
    dragen %>% 
      mutate(is_passed = is_called & is_passed & (is.na(TLOD) | TLOD >= tlod)) %>% 
      count_status() %>% 
      group_by(purity, vartype) %>%
      summarise(
        called = sum(is_called),
        passed = sum(is_passed),
        true = sum(is_true),
        TP = sum(is_tp),
        FP = sum(is_fp),
        FN = sum(is_fn),
        recall = TP / true,
        prec = TP / passed,
        F2 = f_measure(2, prec, recall)
      ) %>% 
      mutate(MinTLOD = tlod)
})
bytlods %>% 
  filter(!is.na(vartype)) %>% 
  mutate(vartype = ifelse(vartype == 1, "SNP", "Indel")) %>% 
  select(MinTLOD, purity, vartype, F2, prec, recall) %>% 
  gather(key = "metric", value = "value", prec, F2, recall) %>% 
  ggplot() +
  geom_line(aes(x = MinTLOD, y = value, color = metric)) + 
  # facet_wrap(~vartype, scales = 'free')
  facet_wrap(vartype~purity, ncol=4, scales = 'free_y')
```

Also exploring NLODs:

```{r}
nlods = c(0:20)
bynlods = nlods %>% map_df(function(nlod) {
  dragen %>% mutate(is_passed = is_called & is_passed & (is.na(NLOD) | NLOD >= nlod)) %>% summarize_stats() %>% mutate(MinNLOD = nlod)
})
bynlods %>% 
  select(MinNLOD, vartype, F2, prec, recall) %>% 
  gather(key = "metric", value = "value", prec, F2, recall) %>% 
  ggplot() +
  geom_line(aes(x = MinNLOD, y = value, color = metric)) + 
  facet_wrap(~vartype, scales = 'free')
```

```{r}
tlod21 = dragen %>% reject_if(TLOD < 21) 
nlod7_15 = dragen %>% reject_if(NLOD < 7 & is_snp | NLOD < 15 & !is_snp)
show_stats(dragen, 
           tlod21,
           tlod21 %>% reject_if(NLOD < 6),
           tlod21 %>% reject_if(NLOD < 7),
           tlod21 %>% reject_if(NLOD < 8)
           )
```

```{r fig.size=3}
# the best values are <20 and <21
filt = data %>% reject_if(caller == 'dragen' & TLOD < 21) %>% count_status() %>% filter(!is_tn)
rbind(data %>% mutate(filtering = 'umccrise'), 
      filt %>% mutate(filtering = 'umccrise+TLOD>=21')
      ) %>% 
  filter(!is_fn) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_grid(filtering~status, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

```{r fig.size=3}
data %>% 
  filter(VD < 100) %>% 
  ggplot() + 
  geom_freqpoly(aes(VD, stat(count), color = caller), binwidth=1) +
  facet_wrap(~status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 100, 10))
```

```{r}
data %>% 
  filter(DP < 200) %>% 
  ggplot() + 
  geom_freqpoly(aes(DP, stat(count), color = caller), binwidth=1) +
  facet_wrap(~status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 200, 20))
```

```{r fig.width=10}
data %>% 
  filter(MQ < 35) %>% 
  ggplot() + 
  geom_freqpoly(aes(MQ, stat(count), color = caller), binwidth=1) +
  facet_wrap(~status, nrow = 3, scales = 'free_y')
```

Filter by MQ perhaps?

```{r fig.width=10}
data %>% 
  filter(MQ > 40) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(~status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

```{r}
show_stats(data %>% filter(caller == 'dragen'), 
           data %>% filter(caller == 'ensemble'),
           
           )
```





















