---
title: "Dragen-vs-bcbio variant evaluation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("evaluation.R")
```

Comparing tumor-normal small somatic variant calling performed by Dragen vs [bcbio-nextgen](https://bcbio-nextgen.readthedocs.org) (BWA, and ensemble 2 out of 3 callers: Strelka2, Mutect2, and Vardict). The input data is COLO829 cell line, a dilution seriues of 40%, 60%, 80%, and 100% tumor purities, made by artificially mixing blood and tumor samples. As a reference truth set, using the one produced by [Craig et al](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4837349/).

Dragen was run using the following arguments:

```
/opt/edico/bin/dragen --output-directory /newvolume/colo100pc_align/ --tumor-fastq1 Colo829_S5_R1_001.fastq.gz --tumor-fastq2 Colo829_S5_R2_001.fastq.gz -1 Colo829_B_S16_R1_001.fastq.gz -2 Colo829_B_S16_R2_001.fastq.gz --output-file-prefix Colo829_100pc --ref-dir /staging/hg38 --enable-variant-caller true --enable-map-align true --vc-sample-name Colo829_100pc --output-format BAM --enable-map-align-output true
```

Variants produced by both pipelines were additionally filtered using the [umccrise approach](https://github.com/umccr/umccrise/blob/master/workflow.md#somatic) (excluding SAGE part).


TODO: try comparing before filtering
TODO: try truth before filtering


```{r}
load_colo = F
load_seqcii = T

load_dragen = T
load_bcbio = T
load_dragen_strelka = F
load_bcbio_strelka = F
dir = "/Users/vsaveliev/Analysis/Dragen/"
```

Load COLO829

```{r}
if (load_colo) {
  truth_vcf  = read.vcf(str_c(dir, "colo/truth_onesample.hg38.PREP.NOALT.ANNO.FILT.vcf.gz"), split.info = T)
  
  bcbio = tibble()
  dragen = tibble()
  dragen_strelka = tibble()
  for (p in c(20, 40, 60, 80, 100)) {
    # bcbio 
    bcbio_callers = c()
    if (load_bcbio)
      bcbio_callers = c(bcbio_callers, "ensemble")
    if (load_bcbio_strelka) 
      bcbio_callers = c(bcbio_callers, "strelka2")
    for (c in bcbio_callers) {
      vcf = read.vcf(str_c(dir, "colo/bcbio_bwa/Colo829_", as.character(p), "pc_", c, "_bwa.vcf.gz"), split.info = T)
      loaded = merge_called_and_truth(vcf, truth_vcf) %>% mutate(caller = str_c("bwa_", c), purity = p)
      bcbio = bind_rows(bcbio, loaded)
    }
    
    # dragen  
    if (load_dragen) {
      vcf = read.vcf(str_c(dir, "colo/dragen/Colo829_", as.character(p), "pc.PREP.NOALT.ANNO.FILT.vcf.gz"), split.info = T)
      loaded =
        merge_called_and_truth(vcf, truth_vcf) %>%
        mutate(TLOD = as.double(TLOD),
               NLOD = as.double(NLOD),
               purity = p,
               caller = "dragen_dragen")
      dragen = bind_rows(dragen, loaded)
    }
    
    # strelka dragen
    if (load_dragen_strelka) {
      vcf = read.vcf(str_c(dir, "colo/strelka/Colo829_", as.character(p), "pc.ANNO.FILT.vcf.gz"), split.info = T)
      loaded = merge_called_and_truth(vcf, truth_vcf) %>% 
        mutate(purity = p,
               caller = "dragen_strelka")
      dragen_strelka = bind_rows(dragen_strelka, loaded)
    }
  }
  
  colo829 = bind_rows(
      dragen_strelka,
      dragen,
      bcbio
    ) %>% mutate(is_true = is_true & FILT.truth == 'PASS') %>% count_status()    
}
```

Reading SEQCII

```{r}
if (load_seqcii) {
  truth_vcf = read.vcf(str_c(dir, "seqcii/truth_SNV_plus_INDEL.ANNO.vcf.gz"), split.info = T) 
  # resque HighConf in FILTER
  truth_vcf$vcf = truth_vcf$vcf %>% mutate(OriginalFilter = FILTER) %>% filter(str_detect(FILTER, "AllPASS;HighConf")) %>% mutate(FILTER = 'PASS')
  truth_vcf$vcf = truth_vcf$vcf %>% mutate(TUMOR_AF = TVAF)
  truth_vcf$vcf = truth_vcf$vcf %>% mutate(NORMAL_AF = NVAF)
  
  bcbio = tibble()
  dragen = tibble()
  dragen_strelka = tibble()
  
  for (p in c(20, 50, 75, 100)) {
    # bcbio
    bcbio_callers = c()
    if (load_bcbio)
      bcbio_callers = c(bcbio_callers, "ensemble")
    if (load_bcbio_strelka) 
      bcbio_callers = c(bcbio_callers, "strelka2")
    for (c in bcbio_callers) {
      vcf = read.vcf(str_c(dir, "seqcii/bcbio/SEQCII_", as.character(p), "pc.vcf.gz"), split.info = T)
      loaded = merge_called_and_truth(vcf, truth_vcf) %>% mutate(caller = str_c("bwa_", c), purity = p)
      bcbio = bind_rows(bcbio, loaded)
    } 
    
    # dragen  
    if (load_dragen) {
      vcf = read.vcf(str_c(dir, "seqcII/dragen/SEQCII_", as.character(p), "pc.PREP.NOALT.ANNO.FILT.vcf.gz"), split.info = T)
      loaded =
        merge_called_and_truth(vcf, truth_vcf) %>%
        mutate(TLOD = as.double(TLOD),
               NLOD = as.double(NLOD),
               purity = p,
               caller = "dragen_dragen")
      dragen = bind_rows(dragen, loaded)
    }
    
    # strelka dragen
    if (load_dragen_strelka) {
      vcf = read.vcf(str_c(dir, "seqcII/dragen_strelka/SEQCII_", as.character(p), "pc.ANNO.FILT.vcf.gz"), split.info = T)
      loaded = merge_called_and_truth(vcf, truth_vcf) %>% 
        mutate(purity = p,
               caller = "dragen_strelka")
      dragen_strelka = bind_rows(dragen_strelka, loaded)
    }
  }
  
  seqcii = bind_rows(
      dragen_strelka,
      dragen,
      bcbio
    ) %>% mutate(is_true = is_true & FILT.truth == 'PASS') %>% count_status()    
}
```

Comparing filtered/unfiltered

```{r}
# mutate(caller = str_c(caller, '_60pc')) %>% 
# data %>% 
#   filter(purity == 60) %>% 
#   filter(is_passed | is_true) %>% 
#   filter(!is.na(vartype)) %>% 
#   ggplot() + 
#   geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
#   facet_wrap(~status, nrow = 3, scales = 'free_y') +
#   scale_x_continuous(breaks = seq(0, 1, 0.1)) +
#   theme_grey(base_size = 16)
# 
# data_unfiltered = data %>% 
#   mutate(
#     umccrise_passed = is_called & AF > 0.1,
#     is_passed = umccrise_passed
#   ) %>%
#   count_status()
# 
# all = bind_rows(data %>% mutate(filtering = 'default'), 
#                 data_unfiltered %>% mutate(filtering = "unfiltered_calls"))
```

```{r}
if (load_seqcii) {
  data = seqcii
} else if (load_colo) {
  data = colo829
}
```

```{r fig.width = 15}
# all = all %>% filter(purity == 60) %>% filter(is_passed | is_true) %>% filter(!is.na(vartype))
# d1 = all %>% filter(filtering == 'default') 
# d2 = all %>% filter(filtering == 'unfiltered_calls')
# library(patchwork)
# p1 = ggplot(d1) + geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) + facet_wrap(~status, nrow = 1, scales = "free") + scale_x_continuous(breaks = seq(0, 1, 0.1))
# p2 = ggplot(d2) + geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) + facet_wrap(~status, nrow = 1, scales = "free") + scale_x_continuous(breaks = seq(0, 1, 0.1)) 
# p1 / p2

#################
## unfiltered with tlods
data_unfiltered = data %>% 
  mutate(
    umccrise_passed = is_called,
    is_passed = umccrise_passed
  ) %>%
  count_status()

data_af10 = data %>% 
  mutate(
    umccrise_passed = is_called & AF >= 0.1,
    is_passed = umccrise_passed
  ) %>%
  count_status()

#################
## compare filtered vs unfiltered
filtered_vs_unfiltered = bind_rows(
  data %>% mutate(filtering = "umccrise"), 
  data_unfiltered %>% mutate(filtering = "no"), 
  data_af10 %>% mutate(filtering = "AF>10%")
)

# filtered_vs_unfiltered %>% 
#   filter(purity == 60) %>% 
#   mutate(caller = factor(caller, levels = c("bwa_ensemble", "dragen_dragen"))) %>% 
#   filter(is_passed | is_true) %>% 
#   ggplot() + 
#   geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
#   facet_wrap(~ status, ncol = 2, scales = 'free_y') +
#   scale_x_continuous(breaks = seq(0, 1, 0.1)) 
```

Summarize filtering reasons for false negatives

```{r fig.width = 8}
bcbio %>% filter(caller == "bwa_ensemble", purity == 20) %>% select(FILT) %>% mutate(filters = str_split(FILT, ';')) %>% mutate(filt = map_chr(filters, ~ifelse(length(.) == 0, "", str_c(sort(as.character(.), decreasing = T), collapse = ';')))) %>% count(filt, sort = T)

bcbio %>% filter(caller == "bwa_ensemble", purity == 20) %>% filter(is_fn) %>% select(FILT) %>% mutate(filters = str_split(FILT, ';')) %>% mutate(filt = map_chr(filters, ~ifelse(length(.) == 0, "", str_c(sort(as.character(.), decreasing = T), collapse = ';')))) %>% count(filt, sort = T)
```


```{r fig.width = 8}
toplot_09 = filtered_vs_unfiltered %>% 
  filter(caller == "bwa_ensemble" | caller == "dragen_dragen") %>%
  mutate(caller = factor(caller, levels = c("bwa_ensemble", "dragen_dragen"))) %>% 
  filter(is_passed | is_true) %>% 
  filter(AF >= 0.1 & AF <= 0.9)

fn = toplot_09 %>% filter(is_fn)
fp = toplot_09 %>% filter(is_fp)
tp = toplot_09 %>% filter(is_tp)
(fn %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), linetype = filtering, color = caller), binwidth = 0.01, alpha = 0.9) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  ylab('False negatives')
) / 
(fp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), linetype = filtering, color = caller), binwidth = 0.01, alpha = 0.9) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  ylab('False positives')
) / 
(tp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), linetype = filtering, color = caller), binwidth = 0.01, alpha = 0.9) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  ylab('True positives')
)
```

Inlcudeing all AF range: from 0% to 100%

```{r fig.width = 8}
toplot = filtered_vs_unfiltered %>% 
  filter(caller != "dragen_dragen_tlod15") %>%
  mutate(caller = factor(caller, levels = c("bwa_ensemble", "dragen_dragen"))) %>% 
  filter(is_passed | is_true)

fn = toplot %>% filter(is_fn)
fp = toplot %>% filter(is_fp)
tp = toplot %>% filter(is_tp)
(fn %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), linetype = filtering, color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0, 1, 0.1)) +
  ylab('False negatives')
) / 
(fp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), linetype = filtering, color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0, 1, 0.1)) +
  ylab('False positives')
) / 
(tp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), linetype = filtering, color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0, 1, 0.1)) +
  ylab('True positives')
)
```

Separately for SNPs and Indels:

```{r fig.width = 8}
toplot = bind_rows(
  data_unfiltered %>% filter(caller == "bwa_ensemble") %>% mutate(caller = 'bcbio: bwa+ensemble'),
  data %>% filter(caller == "bwa_ensemble") %>% mutate(caller = 'bcbio: bwa+ensemble, filters: PoN, gnomAD & AF>10%'),
  # with_tlods_unfiltered %>% filter(caller == "bwa_strelka2") %>% mutate(caller = 'bcbio: bwa+strelka2'),
  data_unfiltered %>% filter(caller == "dragen_dragen") %>% mutate(caller = 'dragen'),
  data %>% filter(caller == "dragen_dragen") %>% mutate(caller = "dragen, filters: PoN, gnomAD & AF>10%")
  # data_unfiltered %>% filter(caller == "dragen_strelka") %>% mutate(caller = "dragen+strelka2"),
  # with_tlods %>% filter(caller == "dragen_dragen_tlod15") %>% mutate(caller = "dragen, filtered with umccrise + tlod>=15")
  # with_tlods %>% filter(caller == "dragen_dragen_tlod21") %>% mutate(caller = "dragen, filtered with umccrise + tlod>=21")
)

toplot = toplot %>% 
  mutate(caller = factor(caller, levels = c("bcbio: bwa+ensemble", 
                                            "bcbio: bwa+ensemble, filters: PoN, gnomAD & AF>10%",
                                            "dragen", 
                                            "dragen, filters: PoN, gnomAD & AF>10%"
                                            # "dragen+strelka2", 
                                            # "dragen, filtered with umccrise + tlod>=15",
                                            # "bcbio: bwa+strelka2"
                                            ))) %>% 
  filter(is_passed | is_true)
  # filter(AF <= 0.9, AF >= 0.1)
  # filter(caller == "dragen+strelka2" | caller == "bcbio: bwa+strelka2")

# toplot2 = toplot %>% 
  # mutate(vartype = factor(ifelse(vartype == 1, "SNP", "Indel"), levels = c("SNP", "Indel")))

# toplot %>% filter(is_tp) %>% 
#   ggplot() + 
#   geom_freqpoly(aes(AF, ..count.., color = caller), binwidth = 0.05) +
#   facet_wrap(vartype ~ purity, nrow = 2, scales = 'free_y') +
#   scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
#   labs(title = "True positives. Left to right: tumor purity percentage")

(toplot %>% filter(is_tp) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, ..count.., color = caller), binwidth = 0.05) +
  facet_wrap(vartype ~ purity, nrow = 2, scales = 'free_y') +
  # scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "True positives. Left to right: tumor purity percentage")
) /
(toplot %>% filter(is_fp) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, ..count.., color = caller), binwidth = 0.05) +
  facet_wrap(vartype ~ purity, nrow = 2, scales = 'free_y') +
  # scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "False positives")
) / 
(toplot %>% filter(is_fn) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, ..count.., color = caller), binwidth = 0.05) +
  facet_wrap(vartype ~ purity, nrow = 2, scales = 'free_y') +
  # scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "False negatives (missed variants)")
) 
```


```{r}
be = with_tlods_unfiltered %>% filter(caller == "bwa_ensemble") %>% mutate(caller = 'bcbio: bwa+ensemble')          %>% filter(purity == 80)
bs =  with_tlods_unfiltered %>% filter(caller == "bwa_strelka2") %>% mutate(caller = 'bcbio: bwa+strelka2')           %>% filter(purity == 80)
d =  with_tlods_unfiltered %>% filter(caller == "dragen_dragen") %>% mutate(caller = 'dragen')                    %>% filter(purity == 80)
du =  with_tlods %>% filter(caller == "dragen_dragen") %>% mutate(caller = "dragen, filtered with umccrise")             %>% filter(purity == 80)
ds =  data_unfiltered %>% filter(caller == "dragen_strelka") %>% mutate(caller = "dragen+strelka2")                    %>% filter(purity == 80)
du15 =  with_tlods %>% filter(caller == "dragen_dragen_tlod15") %>% mutate(caller = "dragen, filtered with umccrise + tlod>=15")%>% filter(purity == 80)
show_stats(bs, ds, d, du, du15, be) 
```

```{r}
dragen_filt = data %>% filter(caller == "dragen_dragen")  # %>%  mutate(vartype = ifelse(vartype == 1, "SNP", "Indel"))
dragen = data_unfiltered %>% filter(caller == "dragen_dragen")  # %>%  mutate(vartype = ifelse(vartype == 1, "SNP", "Indel"))
bcbio_filt = data %>% filter(caller == "bwa_ensemble")  # %>%  mutate(vartype = ifelse(vartype == 1, "SNP", "Indel"))
bcbio = data_unfiltered %>% filter(caller == "bwa_ensemble")  # %>%  mutate(vartype = ifelse(vartype == 1, "SNP", "Indel"))
# p20 = unfiltered %>% filter(purity == 20)
# p60 = unfiltered %>% filter(purity == 50)
# p75 = unfiltered %>% filter(purity == 75)
# p100 = unfiltered %>% filter(purity == 100)
# p20_filt = filtered %>% filter(purity == 20)
# p60_filt = filtered %>% filter(purity == 50)
# p75_filt = filtered %>% filter(purity == 75)
# p100_filt = filtered %>% filter(purity == 100)
dragen_filt_p50 = dragen_filt %>% filter(purity == 50)
dragen_p50 = dragen %>% filter(purity == 50)
bcbio_filt_p50 = bcbio_filt %>% filter(purity == 50)
bcbio_p50 = bcbio %>% filter(purity == 50)
show_stats(bcbio_p50, dragen_p50, bcbio_filt_p50, dragen_filt_p50) 
```

```{r fig.width = 12}
fn = toplot %>% filter(is_fn)
fp = toplot %>% filter(is_fp)
tp = toplot %>% filter(is_tp)
(tp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "True positives. Left to right: tumor purity percentage")
) /
(fp %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "False positives")
) / 
(fn %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth = 0.01) +
  facet_grid(~ purity) +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  labs(title = "False negatives (missed variants)")
) 
```


```{r}
data %>% 
  filter(purity == 60) %>% 
  filter(is_passed | is_true) %>% 
  filter(!is.na(vartype)) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(~ status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0.1, 0.9, 0.1)) +
  theme_grey(base_size = 16)
```

```{r}
data_unfiltered_truthfilt %>% 
  filter(purity == 60) %>% 
  filter(is_passed | is_true) %>% 
  filter(!is.na(vartype)) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(~status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1)) +
  theme_grey(base_size = 16)
```

```{r fig.width=7}
bind_rows(
  dragen,
  dragen %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_tlod21'),
  dragen %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_tlod15')
) %>% 
filter(is_passed | is_true) %>% 
filter(!is.na(vartype)) %>% 
# filter(purity == 60) %>% 
ggplot() + 
geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
facet_wrap(status~purity, ncol=4, scales = 'free_y') +
scale_x_continuous(breaks = seq(0, 1, 0.1))
```

TP charts look suspiciously even for different TLOD thresholds. Checking if it's not a bug by trying few more extreme thresholds. All good:

```{r fix.width=15}
bind_rows(
  dragen,
  dragen %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_tlod21'),
  dragen %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_tlod15'),
  dragen %>% reject_if(TLOD < 30) %>% mutate(caller = 'dragen_tlod30'),
  dragen %>% reject_if(TLOD < 100) %>% mutate(caller = 'dragen_tlod100'),
  dragen %>% reject_if(TLOD < 200) %>% mutate(caller = 'dragen_tlod200')
  ) %>% 
  filter(is_tp) %>% 
  filter(purity == 60) %>%
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

Adding Strelka2. Big piles of FP on low frequency. Perhaps it has to do with Strelka2 not optimized to Dragen alignments. Should try with BWA.

Also wondering about this peak for 100% purity at 100% AF. Is it a problem with T/N calling?

Adding now BWA/ensemble and BWA/strelka.

```{r fig.width=7}
with_tlods = bind_rows(
  dragen %>% mutate(caller = "dragen_dragen"),
  dragen %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_dragen_tlod21'),
  dragen %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_dragen_tlod15'),
  strelka %>% mutate(caller = "dragen_strelka2"),
  data %>% filter(caller == "bwa_strelka2"),
  data %>% filter(caller == "bwa_ensemble")
) 

with_tlods %>% filter(caller == "bwa_ensemble") %>% filter(is_fn) %>% count(AF)

with_tlods %>% 
  mutate(
    caller = factor(caller, levels = c("bwa_ensemble", "dragen_dragen", "dragen_dragen_tlod15", "dragen_dragen_tlod21", "dragen_strelka2", "bwa_strelka2"))
  ) %>% 
  filter(is_passed | is_true) %>% 
  # filter(purity == 60) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(status ~ purity, ncol=4, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

Clearly dragen/dragen and bwa/ensemble are the best combinations. All other mixtures underperform. On on hand, this is expected; on the other, PURPLE and GRIDSS might underperform too running from Dragen alignments.

Now comparing bwa/ensemble vs. dragen/dragen specifically.

```{r fig.width=7}
dragen_bwa = bind_rows(
  dragen %>% mutate(caller = "dragen_dragen"),
  dragen %>% reject_if(TLOD < 21) %>% mutate(caller = 'dragen_dragen_tlod21'),
  dragen %>% reject_if(TLOD < 15) %>% mutate(caller = 'dragen_dragen_tlod15'),
  data %>% filter(caller == "bwa_ensemble")
)
dragen_bwa %>% 
  filter(is_passed | is_true) %>% 
  # filter(purity == 60) %>% 
  # filter(caller == "bwa_ensemble") %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(status~purity, ncol=4, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

Exploring different TLOD thresholds. The threshold around 14 is optimal for purities 60-100, and no filtering is optimal for 40, however 14 looks good anyway.

```{r, fig.width=12}
tlods = c(10:30)
purities = dragen %>% distinct(purity)
bytlods = tlods %>% map_df(function(tlod) {
    # tlod = 5
    dragen %>% 
      mutate(is_passed = is_called & is_passed & (is.na(TLOD) | TLOD >= tlod)) %>% 
      count_status() %>% 
      group_by(purity, vartype) %>%
      summarise(
        called = sum(is_called),
        passed = sum(is_passed),
        true = sum(is_true),
        TP = sum(is_tp),
        FP = sum(is_fp),
        FN = sum(is_fn),
        recall = TP / true,
        prec = TP / passed,
        F2 = f_measure(2, prec, recall)
      ) %>% 
      mutate(MinTLOD = tlod)
})
bytlods %>% 
  filter(!is.na(vartype)) %>% 
  mutate(vartype = ifelse(vartype == 1, "SNP", "Indel")) %>% 
  select(MinTLOD, purity, vartype, F2, prec, recall) %>% 
  gather(key = "metric", value = "value", prec, F2, recall) %>% 
  ggplot() +
  geom_line(aes(x = MinTLOD, y = value, color = metric)) + 
  # facet_wrap(~vartype, scales = 'free')
  facet_wrap(vartype~purity, ncol=4, scales = 'free_y')
```

Also exploring NLODs:

```{r}
nlods = c(0:20)
bynlods = nlods %>% map_df(function(nlod) {
  dragen %>% mutate(is_passed = is_called & is_passed & (is.na(NLOD) | NLOD >= nlod)) %>% summarize_stats() %>% mutate(MinNLOD = nlod)
})
bynlods %>% 
  select(MinNLOD, vartype, F2, prec, recall) %>% 
  gather(key = "metric", value = "value", prec, F2, recall) %>% 
  ggplot() +
  geom_line(aes(x = MinNLOD, y = value, color = metric)) + 
  facet_wrap(~vartype, scales = 'free')
```

```{r}
tlod21 = dragen %>% reject_if(TLOD < 21) 
nlod7_15 = dragen %>% reject_if(NLOD < 7 & is_snp | NLOD < 15 & !is_snp)
show_stats(dragen, 
           tlod21,
           tlod21 %>% reject_if(NLOD < 6),
           tlod21 %>% reject_if(NLOD < 7),
           tlod21 %>% reject_if(NLOD < 8)
           )
```

```{r fig.size=3}
# the best values are <20 and <21
filt = data %>% reject_if(caller == 'dragen' & TLOD < 21) %>% count_status() %>% filter(!is_tn)
rbind(data %>% mutate(filtering = 'umccrise'), 
      filt %>% mutate(filtering = 'umccrise+TLOD>=21')
      ) %>% 
  filter(!is_fn) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_grid(filtering~status, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

```{r fig.size=3}
data %>% 
  filter(VD < 100) %>% 
  ggplot() + 
  geom_freqpoly(aes(VD, stat(count), color = caller), binwidth=1) +
  facet_wrap(~status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 100, 10))
```

```{r}
data %>% 
  filter(DP < 200) %>% 
  ggplot() + 
  geom_freqpoly(aes(DP, stat(count), color = caller), binwidth=1) +
  facet_wrap(~status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 200, 20))
```

```{r fig.width=10}
data %>% 
  filter(MQ < 35) %>% 
  ggplot() + 
  geom_freqpoly(aes(MQ, stat(count), color = caller), binwidth=1) +
  facet_wrap(~status, nrow = 3, scales = 'free_y')
```

Filter by MQ perhaps?

```{r fig.width=10}
data %>% 
  filter(MQ > 40) %>% 
  ggplot() + 
  geom_freqpoly(aes(AF, stat(count), color = caller), binwidth=0.01) +
  facet_wrap(~status, nrow = 3, scales = 'free_y') +
  scale_x_continuous(breaks = seq(0, 1, 0.1))
```

```{r}
show_stats(data %>% filter(caller == 'dragen'), 
           data %>% filter(caller == 'ensemble'),
           
           )
```





















