---
title: "pon"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(knitr)
library(kableExtra)
p <- function(df) { df %>% kable() %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F, position = "float_left") }

source("evaluation.R")
library(patchwork)
library(directlabels)
```

```{r parse_mb}
dir = "~/spa/projects/Saveliev_Somatic_validation/ICGC_MB/"
truth_file = "MB-benchmark.ANNO.FILT.vcf.gz"
called_file = "batch1-ensemble-annotated.ANNO.FILT.vcf.gz"
tumor_sample = "tumor_downsample"

called_vcf = read.vcf(str_c(dir, called_file), split.info = T)
truth_vcf  = read.vcf(str_c(dir, truth_file ), split.info = T)
# mb_truth_vcf$vcf = mb_truth_vcf$vcf %>% filter(str_detect(TIERS, "tier1|tier2"))

mb = merge_called_and_truth(called_vcf, truth_vcf, tumor_sample) %>% 
  mutate(cohort = 'MB',
         sample = cohort)
```

```{r parse_mixture}
dir = "~/rjn/projects/Saveliev_SNV_Filtering/cancer-giab-na12878-na24385/final/20161212_giab-mixture/"
called_file = "24385-ensemble-annotated.ANNO.FILT.vcf.gz"
truth_file = "na12878-na24385-somatic-truth-NORM-ANNO-FILT.vcf.gz"
tumor_sample = "X24385.12878.30.200"

called_vcf = read.vcf(str_c(dir, called_file), split.info = T)
truth_vcf  = read.vcf(str_c(dir, truth_file ), split.info = T, split.samples = T)
truth_vcf$vcf <- truth_vcf$vcf %>% rename(TRUTH_DP = DP)
giab_mix = merge_called_and_truth(called_vcf, truth_vcf, tumor_sample) %>%
  mutate(cohort = "GiaB_mix",
         sample = cohort)
```

```{r parse_colo829}
dir = "~/spa/projects/Saveliev_Somatic_validation/COLO829/"
truth_file = str_c(dir, "truth.NORM.ANNO.FILT.vcf.gz")
truth = read.vcf(truth_file, split.info = T)

vcfs_colo = list( 
  "COLO_40pt"  = "40pc.TP.ANNO.FILT.vcf.gz",
  "COLO_60pt"  = "60pc.ANNO.FILT.vcf.gz",
  "COLO_80pt"  = "80pc.ANNO.FILT.vcf.gz",
  "COLO_100pt" = "100pc.ANNO.FILT.vcf.gz"
) %>% 
  map(~ str_c(dir, .)) %>% 
  map(read.vcf, split.info = T)

colo = vcfs_colo %>% 
  map2(names(vcfs_colo), function(vcf, sample_name) { 
    merge_called_and_truth(vcf, truth) %>% mutate(sample = sample_name)
  }) %>% 
  bind_rows() %>% 
  mutate(cohort = "COLO829",
         sample = cohort)
```

Parsing real patients

```{r patients}
rescue_ponaf = function(.data) {
  .data %>%
    # Dropping the old PoN as we are going to be evaluation the new one
    rescue_filt("PoN") %>%
    # Dropping AF10 filter as well because there are a lot true calls below 10% and we want to test how we call them
    rescue_filt("AF10") %>%
    count_status()
}

# Tothill cohort
dir = "~/spa/data/Results/Tothill-A5/2018-08-11/umccrised/"
vcfs_tothill = list( 
  "PRJ180253_E190.T01.D" = "E190__PRJ180253_E190-T01-D/small_variants/E190__PRJ180253_E190-T01-D-somatic-ensemble.vcf.gz",
  "PRJ180506_E194.T01.D" = "E194__PRJ180506_E194-T01-D/small_variants/E194__PRJ180506_E194-T01-D-somatic-ensemble.vcf.gz",
  "PRJ180494_E199.T01.D" = "E199__PRJ180494_E199-T01-D/small_variants/E199__PRJ180494_E199-T01-D-somatic-ensemble.vcf.gz",
  "PRJ180492_E201.T01.D" = "E201__PRJ180492_E201-T01-D/small_variants/E201__PRJ180492_E201-T01-D-somatic-ensemble.vcf.gz",
  "PRJ180499_E202.T01.D" = "E202__PRJ180499_E202-T01-D/small_variants/E202__PRJ180499_E202-T01-D-somatic-ensemble.vcf.gz" 
) %>% 
  map(~ str_c(dir, .)) %>% 
  map(read.vcf, split.info = T)

data_tothill = vcfs_tothill %>% 
  map2(names(vcfs_tothill), function(vcf, sample_name) { 
    merge_called_and_truth(vcf, NULL, sample_name) %>% mutate(sample = sample_name)
  }) %>% 
  bind_rows() %>% 
  mutate(cohort = "Tothill")

# Highly mutated sample
dir = "~/spa/data/Results/Patients/2019-02-17/SFRC01116/umccrised_newpon/SFRC01116__PRJ190052_SFRC01116T/small_variants/"
tumor_sample = "SFRC01116__PRJ190052_SFRC01116T"
called_vcf = "SFRC01116__PRJ190052_SFRC01116T-somatic-ensemble.vcf.gz"
vcf = str_c(dir, called_vcf) %>% read.vcf(split.info = T)
data_sfrc = merge_called_and_truth(vcf, NULL, tumor_sample) %>% 
  mutate(cohort = "SFRC01116", 
         sample = cohort) %>% 
  filter(!AF < 0.1)

# Bad normal
called_vcf = "~/spa/extras/vlad/pon_analysis/GBB10.NORM.ANNO.FILT.vcf.gz" %>% read.vcf(split.info = T)

data_gbb10 = called_vcf %>% 
  merge_called_and_truth(called_vcf, NULL) %>% 
  mutate(cohort = "GBB10", 
         sample = cohort) %>% 
  filter(!AF < 0.1)
```

Combine samples into single data frame

```{r combine_samples}
data = data_tothill %>% 
  bind_rows(data_sfrc) %>% 
  bind_rows(data_gbb10)
  
data = data %>% 
  # bind_rows(mb %>% filter(is_true) %>% mutate(sample = 'MB_truth')) %>% 
  bind_rows(mb %>% filter(is_true & str_detect(TIERS, "tier1")) %>% mutate(sample = 'MB_truth_tier1')) %>%  # only tier1 is AF>=10%
  # bind_rows(mb %>% filter(is_true & str_detect(TIERS, "tier2")) %>% mutate(sample = 'MB_truth_tier2')) %>% 
  # bind_rows(mb %>% filter(is_true & str_detect(TIERS, "tier3")) %>% mutate(sample = 'MB_truth_tier3')) %>% 
  # bind_rows(mb %>% filter(is_true & str_detect(TIERS, "tier4")) %>% mutate(sample = 'MB_truth_tier4')) %>% 
  bind_rows(mb %>% filter(is_called) %>% mutate(sample = 'MB_called')) %>% 
  bind_rows(mb %>% filter(is_fn) %>% mutate(sample = 'MB_FN'))

data = data %>% 
  bind_rows(giab_mix %>% filter(is_true) %>% mutate(sample = "GiaB_mix_truth")) %>% 
  bind_rows(giab_mix %>% filter(is_called) %>% mutate(sample = 'GiaB_mix_called')) %>% 
  bind_rows(giab_mix %>% filter(is_fn) %>% mutate(sample = 'GiaB_mix_missed'))

data = data %>% 
  bind_rows(colo %>% filter(is_true) %>% mutate(sample = 'COLO829_truth')) %>% 
  bind_rows(colo %>% filter(is_called) %>% mutate(sample = str_c(sample, '_called'))) %>% 
  bind_rows(colo %>% filter(is_fn) %>% mutate(sample = str_c(sample, '_missed'))) 

data_af10 = data %>% filter(!AF < 0.1)

data_af10 %>% count(sample)

data_af10 = data_af10 %>% 
  mutate(
    PoN = ifelse(is.na(PoN), 0, PoN),
    in_gnomAD = str_detect(FILT, 'gnomAD'),
    in_pon = PoN >= 5
  ) 

#data_af10 = data_af10 %>% filter(cohort != "SFRC01116") %>% bind_rows(data_sfrc)

# resc_ponaf = data %>% rescue_ponaf()
# resc_ponafgno = resc_ponaf %>% rescue_filt("gnomAD_common")

# Leaving gnomAD_common from comparison:
# resc_ponaf_nogno = resc_ponaf %>% filter(is.na(FILT) | !str_detect(FILT, "gnomAD_common"))

# Also rescuing called variants for comparison:
# resc_all       = resc_ponaf %>% mutate(is_passed = is_called)
# nogno_resc_all = resc_ponaf_nogno %>% mutate(is_passed = is_called)
```


For each sample, calculate:
- proportion of called variants that were filtered
- proportion of called variants that were filtered by gnomAD
- proportion of called variants that were filtered by PoN
- proportion of good calls fitered by PoN or gnomAD and normal has high enough coverage and low AF

```{r}
fct_case_when <- function(...) {
  args <- as.list(match.call())
  levels <- sapply(args[-1], function(f) f[[3]])  # extract RHS of formula
  levels <- levels[!is.na(levels)]
  factor(dplyr::case_when(...), levels=levels)
}

categorized = data_af10 %>% 
  mutate(
    cohort = ifelse(!is.na(cohort), cohort, sample),
    # expanding Tothill and COLO829 cohorts:
    sample = ifelse(cohort %in% c("Tothill", "COLO829"), sample, cohort)
  ) %>% 
  mutate(  # sorting
    sample = factor(sample, levels = unique(.data$sample))  # %>% fct_relevel("SFRC01116")
  ) %>% 
  mutate(
    bad_normal = NORMAL_DP < 30 | NORMAL_VD >= 1,
    # low_nDP = NORMAL_DP < 30,
    # some_nVD = NORMAL_VD >= 1,
    # purity = 
    # low_freq = TUMOR_AF * purity < 0.3,   # normalize to purity!
    # low_freq = AF < 0.1,
    category = factor(case_when(
      in_pon & in_gnomAD & bad_normal ~ "PoN+gnomAD, normal fail",
      in_pon & in_gnomAD ~ "PoN+gnomAD, normal ok",
      in_pon & bad_normal ~ "PoN, normal fail",
      in_pon ~ "PoN, normal ok",
      in_gnomAD & bad_normal ~ "gnomAD, normal fail",
      in_gnomAD ~ "gnomAD, normal ok",
      AF < 0.1 ~ "AF<10%", 
      !umccrise_passed ~ "other filtered",
      TRUE ~ "passed"
    ), levels = c("PoN, normal fail", "PoN, normal ok", 
                  "PoN+gnomAD, normal fail", "PoN+gnomAD, normal ok",
                  "gnomAD, normal fail", "gnomAD, normal ok",
                  "other filtered", "passed"))
  )

# categorized %>% filter(sample == "SFRC01116") %>% count(in_gnomAD)
# stats %>% filter(sample == "SFRC01116") %>% filter(category != "passed")

all_stats = categorized %>% 
  filter(!str_detect(sample, "COLO"), 
         !str_detect(sample, "GiaB_mix")) %>% 
  group_by(vartype, sample) %>% 
  mutate(total_vars = n()) %>% 
  ungroup() %>% 
  group_by(vartype, sample, category) %>% 
  summarize(
    cnt = n(),
    total_vars = mean(total_vars),
    share = cnt / total_vars * 100
  )

reject_only_stats = categorized %>% 
  filter(!str_detect(sample, "COLO"), 
         !str_detect(sample, "GiaB_mix")) %>% 
  filter(category != 'passed') %>% 
  group_by(vartype, sample) %>% 
  mutate(total_vars = n()) %>% 
  ungroup() %>% 
  group_by(vartype, sample, category) %>% 
  summarize(
    cnt = n(),
    total_vars = mean(total_vars),
    share = cnt / total_vars
  )
```

```{r fig.width=15}
data_af10 %>% 
  ggplot() +
  geom_violin(aes(sample, AF)) +
  coord_flip() +
  ylim(0, 1)

?geom_boxplot
```

```{r fig.width=25}
all_stats %>% 
  # filter(vartype == "SNP") %>% 
  ggplot(aes(sample, share, group = forcats::fct_rev(category))) + 
  geom_bar(aes(fill = category), stat = "identity", position = "stack") +
  geom_text(aes(label = cnt), check_overlap = T, position = position_stack(vjust = 0.5), size = 2.5) +
  # geom_text(aes(y = -2, label = total_vars), hjust = "right", size = 2.5) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  # ylim(-12, 101) +
  coord_flip() +
  facet_wrap(~vartype, nrow = 2)
```

```{r indels}
all_stats %>% 
  filter(vartype != "SNP") %>% 
  ggplot(aes(sample, share, group = category)) + 
  geom_bar(aes(fill = category), stat = "identity", position = "stack") +
  geom_text(aes(label = cnt), check_overlap = T, position = position_stack(vjust = 0.5), size = 2.5) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  coord_flip()
```

```{r filtered_only}
reject_only_stats %>% 
  filter(vartype == "SNP") %>% 
  filter(!str_detect(sample, "COLO"), 
         !str_detect(sample, "GiaB_mix")
  ) %>% 
  ggplot(aes(sample, share, group = category)) + 
  geom_bar(aes(fill = category), stat = "identity", position = "stack") +
  geom_text(aes(label = cnt), check_overlap = T, position = position_stack(vjust = 0.5), size = 2.5) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  coord_flip()
```

```{r filtered_only_indels}
reject_only_stats %>% 
  filter(vartype != "SNP") %>% 
  ggplot(aes(sample, share, group = category)) + 
  geom_bar(aes(fill = category), stat = "identity", position = "stack") +
  geom_text(aes(label = cnt), check_overlap = T, position = position_stack(vjust = 0.5), size = 2.5) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  coord_flip()
```

We see that:
* for the 5 patients, PoN is mostly helpful in sites with a low normal coverage, which is good and exprected.
* for SFRC01116, it's a bit the other way around: very few PoN are in a low normal coverage.
* ICGC MB benchmark PoN is mostly in fine normal regions.
* The GiaB mixture is also mostly in fine normal regions.

Build histogram (or even box plots?) of average NORMAL_VD and NORMAL_AF in PoN and outside of PoN.

```{r fig.height=5, fig.width=10}
boxplot_data = data_af10 %>% 
  filter(sample != "GBB10") %>% 
  mutate(
    cohort = ifelse(!is.na(cohort), cohort, sample),
    # expanding Tothill and COLO829 cohorts:
    sample = ifelse(cohort %in% c("Tothill", "COLO829", "MB"), sample, cohort),
    is_benchmark = cohort %in% c("MB", "COLO829")
  ) %>% 
  filter(sample != "GiaB_mix_truth", 
         sample != "GiaB_mix_missed", 
         !str_detect(sample, "missed"), 
         !str_detect(sample, "FN")
         )
(boxplot_data %>% 
  filter(!is_benchmark) %>% 
  ggplot() +
  scale_colour_manual(values = c("dark green", "red")) +
  geom_boxplot(aes(sample, NORMAL_DP, color = in_pon)) +
  coord_flip() +
  ylim(0, 100)
) / 
(boxplot_data %>% 
  filter(is_benchmark) %>% 
  ggplot() +
  scale_colour_manual(values = c("dark green", "red")) +
  geom_boxplot(aes(sample, NORMAL_DP, color = in_pon)) +
  coord_flip() +
  ylim(0, 100)
) + plot_layout(ncol = 1, heights = c(3, 1))
```

Define for each sample a metric on % filtered with PoN and gnomAD in good normal:

```{r}
data_af10 %>% 
  mutate(
    cohort = ifelse(!is.na(cohort), cohort, sample),
    sample = ifelse(cohort == "Tothill" , sample, cohort),  # explanding Tothill cohort
    good_normal = NORMAL_DP > 30 & NORMAL_VD <= 1
  ) %>% 
  group_by(vartype, sample) %>% 
  summarize(
    pct_pon_in_good_normal = mean(in_pon & good_normal, rm.na = T) / mean(in_pon, rm.na = T)
  ) %>% 
  ggplot() +
  geom_col(aes(x = sample, y = pct_pon_in_good_normal, fill = vartype), position = "dodge") +
  coord_flip()
```



### Comparing old and new PoN

Parsing entire PoN [don't do - too many variants]

```{r}
pons_snps = list(
  "T-only Mutect2" = "~/spa/extras/vlad/synced/umccr/vcf_stuff/vcf_stuff/panel_of_normals/pon_Feb26/GRCh37/pon.snps.vcf.gz",
  "Germline GATK" = "~/spa/extras/vlad/synced/umccr/vcf_stuff/vcf_stuff/panel_of_normals/pon_gatk/GRCh37/pon.snps.vcf.gz",
  "Germline ensemble" = "~/spa/extras/vlad/synced/umccr/vcf_stuff/vcf_stuff/panel_of_normals/pon_ensemble/GRCh37/pon.snps.vcf.gz"
) %>% map(~ .[])
pons_indels = list(
  "T-only Mutect2" = "~/spa/extras/vlad/synced/umccr/vcf_stuff/vcf_stuff/panel_of_normals/pon_Feb26/GRCh37/pon.indels.vcf.gz",
  "Germline GATK" = "~/spa/extras/vlad/synced/umccr/vcf_stuff/vcf_stuff/panel_of_normals/pon_gatk/GRCh37/pon.indels.vcf.gz",
  "Germline ensemble" = "~/spa/extras/vlad/synced/umccr/vcf_stuff/vcf_stuff/panel_of_normals/pon_ensemble/GRCh37/pon.indels.vcf.gz"
)
```


Parse MB annotated with the recalled PoN, as well as built from ensemble and gatk germline calls.

```{r}
dir = "~/spa/projects/Saveliev_Somatic_validation/ICGC_MB/"
truth_file = "MB-benchmark.ANNO.FILT.PON_ens.PON_gakt.vcf.gz"
called_file = "batch1-ensemble-annotated.ANNO.FILT.PON_ens.PON_gatk.vcf.gz"
tumor_sample = "tumor_downsample"

called_vcf = read.vcf(str_c(dir, called_file), split.info = T)
truth_vcf  = read.vcf(str_c(dir, truth_file ), split.info = T, split.samples = T)
# mb_truth_vcf$vcf = mb_truth_vcf$vcf %>% filter(str_detect(TIERS, "tier1|tier2"))

mb = merge_called_and_truth(called_vcf, truth_vcf, tumor_sample) %>% mutate(sample = 'MB')
mb_pons = mb %>%
  set_tumor_field("PoN") %>%
  set_tumor_field("PoN_ens") %>%
  set_tumor_field("PoN_gatk") %>%
  mutate(
    PoN = ifelse(is.na(PoN), 0, PoN),
    PoN_ens = ifelse(is.na(PoN_ens), 0, PoN_ens),
    PoN_gatk = ifelse(is.na(PoN_gatk), 0, PoN_gatk),
    in_gnomAD = str_detect(FILT, 'gnomAD'),
    pon2 = PoN >= 2,
    pon5 = PoN >= 5,
    pon_ens2 = PoN_ens >= 2,
    pon_gatk2 = PoN_gatk >= 2,
    pon_ens5 = PoN_ens >= 5,
    pon_gatk5 = PoN_gatk >= 5,
    in_pon = pon5
  )
```


```{r hit_distribution}
mb_pons %>% 
  ggplot() + 
  geom_freqpoly(aes(PoN), binwidth = 1) + 
  geom_freqpoly(aes(PoN_ens), binwidth = 1, color = 'red') +
  geom_freqpoly(aes(PoN_gatk), binwidth = 1, color = 'blue') +
  ylim(0, 100)
```

The new panel of normals has a shape of 2 distributions: spike at low counts (1-5), and a hill at ~115 (which is 50% of total sample counts). Wondering what is the nature of the variants making this bell shaped normal distribution?

```{r hit_distribution_no_gnomad}
mb_pons %>% filter(!str_detect(FILT, "gnomAD_common")) %>% 
  ggplot() + 
  geom_freqpoly(aes(PoN), binwidth = 1) + 
  geom_freqpoly(aes(PoN_ens), binwidth = 1, color = 'red') +
  geom_freqpoly(aes(PoN_gatk), binwidth = 1, color = 'blue') +
  ylim(0, 100)
```

Though not so clear bell shape if we remove gnomAD variants. Now a step-shaped plot, with a flat line until 50% of total samples, then a lower level.

Hypothesis: gnomAD is helpful enough for the MB benchmark?

```{r new_pon}
resc_ponaf = mb_pons %>% rescue_ponaf()
resc_gno = resc_ponaf %>% rescue_filt("gnomAD_common")
nogno = resc_ponaf %>% filter(!str_detect(FILT, "gnomAD_common"))

# plot x axis - PoN cnt

resc_ponaf_new1 = resc_ponaf %>% reject_if(PoN >= 1)
resc_ponaf_new2 = resc_ponaf %>% reject_if(PoN >= 2)
resc_ponaf_new5 = resc_ponaf %>% reject_if(PoN >= 5)
resc_gno_new1 = resc_gno %>% reject_if(PoN >= 1)
resc_gno_new2 = resc_gno %>% reject_if(PoN >= 2)
resc_gno_new5 = resc_gno %>% reject_if(PoN >= 5)
nogno_new1 = nogno %>% reject_if(PoN >= 1)
nogno_new2 = nogno %>% reject_if(PoN >= 2)
nogno_new5 = nogno %>% reject_if(PoN >= 5)

resc_ponaf_ens1 = resc_ponaf %>% reject_if(PoN_ens >= 1)
resc_ponaf_ens2 = resc_ponaf %>% reject_if(PoN_ens >= 2)
resc_ponaf_ens5 = resc_ponaf %>% reject_if(PoN_ens >= 5)
resc_gno_ens1 = resc_gno %>% reject_if(PoN_ens >= 1)
resc_gno_ens2 = resc_gno %>% reject_if(PoN_ens >= 2)
resc_gno_ens5 = resc_gno %>% reject_if(PoN_ens >= 5)
nogno_ens1 = nogno %>% reject_if(PoN_ens >= 1)
nogno_ens2 = nogno %>% reject_if(PoN_ens >= 2)
nogno_ens5 = nogno %>% reject_if(PoN_ens >= 5)

resc_ponaf_gatk1 = resc_ponaf %>% reject_if(PoN_gatk >= 1)
resc_ponaf_gatk2 = resc_ponaf %>% reject_if(PoN_gatk >= 2)
resc_ponaf_gatk5 = resc_ponaf %>% reject_if(PoN_gatk >= 5)
resc_gno_gatk1 = resc_gno %>% reject_if(PoN_gatk >= 1)
resc_gno_gatk2 = resc_gno %>% reject_if(PoN_gatk >= 2)
resc_gno_gatk5 = resc_gno %>% reject_if(PoN_gatk >= 5)
nogno_gatk1 = nogno %>% reject_if(PoN_gatk >= 1)
nogno_gatk2 = nogno %>% reject_if(PoN_gatk >= 2)
nogno_gatk5 = nogno %>% reject_if(PoN_gatk >= 5)

show_stats(
  resc_ponaf,
  resc_ponaf_new5,
  resc_ponaf_ens2,
  resc_ponaf_gatk2,
  resc_gno,
  resc_gno_new5,
  resc_gno_ens2,
  resc_gno_gatk2,
  nogno,
  nogno_new5,
  nogno_ens2,
  nogno_gatk2
) %>% 
  filter(!is.na(vartype))
```

We see that PoN and gnomAD are complementary to each other and PoN always add something to F2 regardless of gnomAD.

Also we see that neither don't help much with MB benchmark.


Hypothesis: PoN should have the highest effect in the areas of gaps of coverage in normals.

```{r normal_depth}
ggplot() + 
  geom_density(data = resc_ponaf %>% filter(str_detect(FILT, 'gnomAD')), 
               aes(NORMAL_DP), color = 'red') +
  geom_density(data = resc_ponaf %>% filter(PoN >= 5),
               aes(NORMAL_DP), color = 'blue') +
  geom_density(data = resc_ponaf %>% filter(PoN_ens >= 2),
               aes(NORMAL_DP), color = 'green') +
  geom_density(data = resc_ponaf %>% filter(!str_detect(FILT, 'gnomAD'), !PoN >= 5, !PoN_ens >= 2), 
               aes(NORMAL_DP), color = 'black')
```

Some correlation between presence in PoN/gnomAD and normal depth.

Exploring variants in PoN/gnomAD with high normal depth in detail.

```{r normal_depth}
resc_ponaf %>% 
  filter(is_snp, str_detect(FILT, 'gnomAD') | PoN >= 5) %>% 
  filter(is_true) %>%
  filter(NORMAL_DP > 30) %>%
  # filter(NORMAL_VD == 0) %>%
  mutate(
    AF = format(AF * 100, digits = 3),
    NORMAL_AF = format(NORMAL_AF * 100, digits = 3)
  ) %>% 
  select(PCGR_TIER, GENE, VD, AF, DP, MQ, NORMAL_DP, NORMAL_AF, NORMAL_VD, NORMAL_MQ, CHROM, POS, REF, ALT, PoN_ens, PoN_gatk, PoN, FILT, CALLS) %>% p
```

47 gnomAD or PoN variants have no support is normal match. What are they?

TODO: report germline leakage in MultiQC or Rmd? gnomAD and germline PoN (old-style PoN) with high enough normal depth?

How many variants are unique to the new PoN or the old PoN:

```{r}
# The new PoN is missing 9:
resc_ponaf %>% filter(PoN_gatk >= 5 & PoN < 5) %>% nrow

# The gatk PoN is missing 543:
resc_ponaf %>% filter(PoN >= 5 & PoN_gatk < 5) %>% nrow

# The ensemble PoN is missing 615:
resc_ponaf %>% filter(PoN >= 5 & PoN_ens < 5) %>% nrow

# The new PoN is missing 35:
resc_ponaf %>% filter(PoN_gatk >= 2 & PoN < 5) %>% nrow

# The gatk PoN is missing 248:
resc_ponaf %>% filter(PoN >= 5 & PoN_gatk < 2) %>% nrow

# The ensemble PoN is missing 295:
resc_ponaf %>% filter(PoN >= 5 & PoN_ens < 2) %>% nrow
```

Determining the best threshold.

```{r, fig.width = 5}
stats = tibble(hits = integer(0),
               metric = character(0),
               pon_type = character(0),
               val = double(0)
               )

cutoffs = c(1:9, seq(10, max(resc_ponaf$PoN) + 19, by = 20))

for (n in cutoffs) {
  for (metric in c("F2", "prec", "recall")) {
    for (pon_type in c("PoN", "PoN_ens")) {
      val = resc_ponaf %>% reject_if(eval(sym(pon_type)) >= n) %>% summarize_stats() %>% filter(vartype == "SNP") %>% .[[metric]]
      stats = stats %>% add_row(
        hits = n,
        metric = metric,
        pon_type = pon_type,
        val = val
      )
    }
  }
}

stats = stats %>% 
  mutate(
    label = ifelse(pon_type == "PoN", "Tumor only", "Germline"),
    scale = ifelse(hits >= 10, "hits >= 10", "hits < 10")
  )

library(directlabels)
dl_method = list(dl.trans(x = x + 0.2), "maxvar.points", cex = 0.8)
# nopon_stats = resc_ponaf %>% summarize_stats() %>% filter(vartype == "SNP")
stats %>% 
  ggplot(aes(x = hits, y = val, color = label, linetype = metric)) + 
  geom_line() +
  # scale_colour_discrete(guide = 'none') +
  scale_x_continuous(breaks = c(0, cutoffs)) +  #, expand = expand_scale(add = c(1, 5))) +
  # geom_dl(aes(y = val, label = str_c(metric, ", ", label)), method = dl_method) +
  # geom_dl(aes(y = recall_snps, label = str_c("Recall, ", label)), method = geom_dl_m) +
  # geom_dl(aes(y = prec_snps, label = str_c("Prec, ", label)), method = geom_dl_m) +
  # geom_hline(yintercept = resc_ponaf %>% summarize_stats() %>% filter(vartype == "SNP") %>% .$F2,
  #            color = label) +
  # geom_hline(yintercept = resc_ponaf %>% summarize_stats() %>% filter(vartype == "SNP") %>% .$recall,
  #            color = label) +
  facet_wrap(~scale, nrow = 1, scales = "free_x") +
  ylim(0.82, 0.95)
 # +
 #  geom_line(aes(y = nopon_recall)) +
 #  geom_line(aes(y = nopon_prec))
  
  # facet_wrap(~vartype, nrow = 2, scales = "free_y")
```

```{r}
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

stats %>% 
  filter(scale == "hits < 10") %>%  
  ggplot(aes(x = hits, y = val, linetype = metric)) + 
  geom_line(color = gg_color_hue(2)[[2]]) +
  scale_x_continuous(breaks = c(0, cutoffs), expand = expand_scale(add = c(0.1, 1))) +
  geom_dl(aes(y = val, label = metric), method = dl_method)
```

The best threshold for SNPs is 5 for the new PoN. For the old ones, it's 2.
However, in terms of indels, the best is not to filter at all: probably the truth set has too many false positives.



Other benchmarks to determine the best threshold:

```{r, fig.width = 7}
benchmarks = bind_rows(mb, colo, giab_mix) %>% rescue_ponaf()
b2 = benchmarks %>% rescue_ponaf()

bench_stats = tibble(
  hits = integer(0),
  metric = character(0),
  cohort = character(0),
  val = double(0)
)

for (n in cutoffs) {
  for (metric in c("F2", "prec", "recall")) {
    for (coh in c("MB", "COLO829", "GiaB_mix")) {
      val = b2 %>% filter(cohort == coh) %>% reject_if(PoN >= n) %>% summarize_stats() %>% filter(vartype == "SNP") %>% .[[metric]]
      bench_stats = bench_stats %>% add_row(
        hits = n,
        metric = metric,
        cohort = coh,
        val = val
      )
    }
  }
}

library(directlabels)
geom_dl_m = list(dl.trans(x = x + 0.2), "last.points", cex = 0.8)

bench_stats %>% 
  mutate(
    scale = ifelse(hits >= 10, "hits >= 10", "hits < 10")
  ) %>% 
  ggplot(aes(x = hits, y = val, color = cohort, linetype = metric)) + 
  ylab("3 benhcmarks, 3 metirc values for each") +
  xlab("X: PoN hits (correspnds to variants after filtering those that hit the panel at least X times)") +
  geom_line() +
  geom_dl(aes(y = val, label = metric), method = geom_dl_m) +
  scale_x_continuous(breaks = c(0, cutoffs), expand = expand_scale(add = c(0.1, 1))) +
  # ylim(0.8, 0.95) +
  facet_grid(cohort ~ scale, scales = "free")

```

Plotting NORMAL_DP and NORMAL_VD distributions for this in gnomAD, for those in PoN, and others.

GnomAD and PoN variants have on avareage a lower depth in normal, as well as AF:

```{r}
mb_called = resc_ponaf %>% filter(is_called)
# (mb_called %>% 
#   ggplot() +
#   scale_colour_manual(values = c("blue", "orange")) +
#   geom_density(aes(NORMAL_AF, color = in_gnomAD)) +
#   xlim(0, 0.05)
# ) /
# (mb_called %>% 
#   ggplot() +
#   geom_density(aes(NORMAL_DP, color = in_gnomAD)) +
#   scale_colour_manual(values = c("blue", "orange")) +
#   xlim(0, 100)
# ) /
  
(mb_called %>% 
  mutate(`In PoN` = in_pon) %>% 
  ggplot() +
  scale_colour_manual(values = c("blue", "orange")) +
  geom_density(aes(NORMAL_AF, color = `In PoN`)) +
  xlim(0, 0.05)
) /
(mb_called %>% 
  mutate(`In PoN` = in_pon) %>% 
  ggplot() +
  scale_colour_manual(values = c("blue", "orange")) +
  geom_density(aes(NORMAL_DP, color = `In PoN`)) +
  xlim(0, 100)
)
```

```{r}
mb_called %>% 
  group_by(in_gnomAD, in_pon) %>% 
  mutate(NORMAL_VD = NORMAL_DP * NORMAL_AF) %>% 
  summarize(
    mean_normal_dp = format(mean(NORMAL_DP, na.rm = T), digits = 3),
    mean_normal_vd = format(mean(NORMAL_VD, na.rm = T), digits = 3),
    mean_normal_af = format(percent(mean(NORMAL_AF, na.rm = T)), digits = 3)
  )
```

Same excersize for the patients:

```{r}
patients = data_af10 %>% filter(cohort == "Tothill")
(patients %>% 
  ggplot() +
  geom_density(aes(NORMAL_AF, color = in_gnomAD)) +
  xlim(0, 0.05)
) /
(patients %>% 
  ggplot() +
  geom_density(aes(NORMAL_DP, color = in_gnomAD)) +
  xlim(0, 100)
) /
(patients %>% 
  ggplot() +
  geom_density(aes(NORMAL_AF, color = in_pon)) +
  xlim(0, 0.05)
) /
(patients %>% 
  ggplot() +
  geom_density(aes(NORMAL_DP, color = in_pon)) +
  xlim(0, 100)
)
```

```{r}
data_af10 %>% 
  filter(is_passed) %>% 
  group_by(in_pon) %>% 
  summarize(mean_normal_af = percent(mean(NORMAL_AF, rm.na = T)))
```








