# Snakemake file for filtering of a VCF file against a set of normal samples.

# Usage: 
# snakemake -p --configfile=config.yaml

import os
from os.path import isfile
import glob
import re
from python_utils.hpc import get_ref_file


# configfile: 'config.yaml'

ref_fasta = config.get('ref_fa') or get_ref_file(config['genome'])
normals_dir = config.get('normals_dir') or get_ref_file(config['genome'], key='panel_of_normals_dir')

hits_thresholds = config.get('hits_thresholds', [1, 2, 3])
samples = config.get('samples')
if isinstance(samples, str):
    samples = {'sample': samples}


rule all:
    input:
        expand('pon_filter/{sample}-n{ht}.vcf.gz', sample=samples, ht=hits_thresholds)


rule prep_vcf:
    input:
        lambda wildcards: samples[wildcards.sample]
    output:
        'pon_prep_vcf/{sample}.vcf.gz'
    shell:
        'bcftools view {input} -f.,PASS -Oz -o {output}'


rule pon_anno:
    input:
        vcf = rules.prep_vcf.output
    output:
        'pon_filter/{sample}-n{ht}.vcf.gz'
    params:
        ht = lambda wildcards: wildcards.ht
    shell:
        'pon_anno {input.vcf} -h {params.ht} | bcftools view -f.,PASS -Oz -o {output} && tabix -p vcf {output}'
